File: Projects/ODEsWithSeasonalities/Mathematica/AirPollutionODEsSolverInterface.m
(* Air Pollution ODE Solver Interface Mathematica Package *)

(* :Title: AirPollutionODEsSolverInterface *)
(* :Context: AirPollutionODEsSolverInterface` *)
(* :Author: Anton Antonov *)
(* :Date: 2015-10-22 *)

(* :Package Version: 0.1 *)
(* :Mathematica Version: *)
(* :Copyright: (c) 2015 Anton Antonov *)
(* :Keywords: air pollution, ODE, interface *)
(* :Discussion:

 For the background on this numerical simulation interface see the discussion

   "ODE w/seasonal forcing term",

   http://mathematica.stackexchange.com/questions/95015/ode-w-seasonal-forcing-term/

 This file was created with Mathematica Plugin for IntelliJ IDEA.

 Anton Antonov
*)

V = 28*10^6;
Manipulate[
  DynamicModule[{fsols, c, F, Cin},
    F[t_] := 10^6 (1 + 6*Sin[2 \[Pi] t]);
    Cin[t_] := 10^6*(10 + 10*Cos[2 \[Pi] t]);
    fsols =
        Table[Block[{sol},
          F[t_] := 10^6*(1 + 6.0*Sin[2 \[Pi] t]);
          Cin[t_] := 10^6 (10 + 10*Cos[2 \[Pi] t]);
          sol =
              NDSolve[{c'[t] == m*F[t]/V (Cin[t] - c[t]), c[0] == k*10^7.},
                c[t], {t, 0, tEnd}, Method -> Automatic];
          c[t] /. sol[[1]]
        ], {k, kMin, kMax, 0.05}];
    Plot[fsols/10^6, {t, 0, tEnd}, PlotRange -> {All, All}, AspectRatio -> 1/2]
  ],
  {{m, 6, "RHS factor"}, 0., 15, 0.5},
  {{kMin, 0, "min initial condition factor"}, 0, 2, 0.01},
  {{kMax, 0.6, "max initial condition factor"}, 0, 2, 0.01},
  {{tEnd, 8, "time interval (years)"}, 1, 20, 0.5}]

File: Projects/ODEsWithSeasonalities/R/AirPollutionODEsSolverInterface.R
##---
## Title: Air Pollution ODE's Solver Interface
## Author: Anton Antonov
## Start date: 2015-09-29
##---

library(shiny)
library(deSolve)

Pi <- 3.14159265
V <- 28.0 * 10.0^6;

Fr <- function(t) { 10.0^6 * ( 1.0 + 6.0 * sin(2*Pi*t) ) }
Cin <- function(t) { 10.0^6 * ( 10.0 + 10.0 * cos(2*Pi*t) ) }

PFunc <- function( t, y, m ) { list( m[1] * Fr(t) / V * ( Cin(t) - y[1] ) ) } 

server <- function(input, output) {
  
  output$solutionPlot <- renderPlot({
    if ( input$kmin < input$kmax) {
      for( k in seq( input$kmin, input$kmax, 0.05 ) ) {
        yini <- c( y1 = k*10^7. )
        ysol <- ode( y = yini, func = PFunc, times = seq( 0, input$tend, 0.01 ), parms = input$m, method = "ode45" )
        if ( k == input$kmin )  {
          plot( ysol, type = "l", which = "y1", lwd = 2, ylab = "concentration", main = "", ylim = c( 0, 1 * input$kmax * Cin(0) ) )
        } else { 
          lines( ysol, type = "l", lwd = 2 )
        }
      }
    }
  })
}

ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(
      sliderInput("m", "RHS factor:", min = 0, max = 15, step = 0.5, value = 6.0 ),
      sliderInput("kmin", "min initial condition factor:", min = 0, max = 2, step = 0.01, value = 0 ),
      sliderInput("kmax", "max initial condition factor:", min = 0, max = 2, step = 0.01, value = 0.6 ),
      sliderInput("tend", "time interval (years):", min = 1, max = 10, step = 0.5, value = 8 )      
    ),
    mainPanel( plotOutput("solutionPlot") )
  )
)

shinyApp(ui = ui, server = server)
